### PR TITLE
docs/user/vsphere: Add static IP via Afterburn info

### PR DESCRIPTION
**Pull request by cgwalters**
I didn't test this myself, but we should have landed the work
for 4.6.

### CODE REVIEW

**Comment by cgwalters** [comment]
Also installer team: We should think about a way to use this in IPI scenarios.  Anything of this form will require something like extending install-config with "per node configuration" which also gets into https://github.com/openshift/machine-config-operator/issues/1720

File: docs/user/vsphere/install_upi.md [hunk 1]
```diff
@@ -255,6 +255,26 @@ TODO RHEL CoreOS does not have assets for vSphere.
 
 Set the vApp properties of the VM to set the Ignition config for the VM. The `guestinfo.ignition.config.data` property is the base64-encoded Ignition config. The `guestinfo.ignition.config.data.encoding` should be set to `base64`.
 
+### Configuring RHCOS VMs with static IP addresses
+
+As of OpenShift 4.6, you may use a new `guestinfo.afterburn.initrd.network-kargs` to configure static IP addressing.  
+The CoreOS "afterburn" component will lookup for the well-known key `guestinfo.afterburn.initrd.network-kargs` and use its value instead of the default.
```
Comment (abhinavdahiya) [inline docs/user/vsphere/install_upi.md:261]: So this only sets the first boot and what about after that??

Should we show a complete example here?

Comment (cgwalters) [reply docs/user/vsphere/install_upi.md:261]: Hmm you're right, we may need to document here how to also embed the static addressing in the Ignition config to persist for subsequent boots.  I will get back to you on this.

Comment (jcpowermac) [reply docs/user/vsphere/install_upi.md:261]: Hmm that is a good point what if I have to re-address my RHCOS virtual machine?

**Comment by jcpowermac** [comment]
The PR seems fine to me as far as documentation. My only concern is the confusion between our current terraform ip addressing and the addition of the extra config.

Was static addressing a feature for 4.6? I only see this possible for 4.6 in UPI with documentation change but I doubt QE has tested it. I also think we should be updating the UPI terraform template to replace our current method of addressing (which is different than #3533). And is something we should already be testing in CI for 4.6?

> Also installer team: We should think about a way to use this in IPI scenarios. Anything of this form will require something like extending install-config with "per node configuration" which also gets into [openshift/machine-config-operator#1720](https://github.com/openshift/machine-config-operator/issues/1720)

Based that is a guest's extra config I would think MAO be a better location at least for compute. 
https://github.com/openshift/machine-api-operator/blob/master/pkg/controller/vsphere/reconciler.go#L41-L45
The installer could set it for control plane.

**Comment by abhinavdahiya** [comment]
> The PR seems fine to me as far as documentation. My only concern is the confusion between our current terraform ip addressing and the addition of the extra config.
> 
> Was static addressing a feature for 4.6? I only see this possible for 4.6 in UPI with documentation change but I doubt QE has tested it. I also think we should be updating the UPI terraform template to replace our current method of addressing (which is different than #3533). And is something we should already be testing in CI for 4.6?
> 
> > Also installer team: We should think about a way to use this in IPI scenarios. Anything of this form will require something like extending install-config with "per node configuration" which also gets into [openshift/machine-config-operator#1720](https://github.com/openshift/machine-config-operator/issues/1720)

"per node configuration" is something I'm not on board at all for configuration using IPI, IMO that goes against what we want the experience to look like for self-managing autoscaling default experience for IPI clusters. If we can look at how to allow users to provider the installer will a range of IP addresses that the cluster can use to add static ip to one of the Machine created that is something I'm more inclined towards as next steps.

> 
> Based that is a guest's extra config I would think MAO be a better location at least for compute.
> https://github.com/openshift/machine-api-operator/blob/master/pkg/controller/vsphere/reconciler.go#L41-L45
> The installer could set it for control plane.

**Comment by jcpowermac** [comment]
> "per node configuration" is something I'm not on board at all for configuration using IPI, IMO that goes against what we want the experience to look like for self-managing autoscaling default experience for IPI clusters. If we can look at how to allow users to provider the installer will a range of IP addresses that the cluster can use to add static ip to one of the Machine created that is something I'm more inclined towards as next steps.

Completely agree. I am hoping this operator that metal uses might help with that but it is certainly not something that I have had a chance to investigate:
https://github.com/metal3-io/ip-address-manager

**Comment by cgwalters** [comment]
> "per node configuration" is something I'm not on board at all for configuration using IPI, IMO that goes against what we want the experience to look like for self-managing autoscaling default experience for IPI clusters. If we can look at how to allow users to provider the installer will a range of IP addresses that the cluster can use to add static ip to one of the Machine created that is something I'm more inclined towards as next steps.

Yeah...I think we may need to create a term like "external-installer-automated-provisioned-infrastructure" (extIPI? dunno) - basically people who are *automating* UPI installs (much like we are in CI for UPI).  In some scenarios it's OK not to have autoscaling/machineAPI.  You're probably right that extIPI wouldn't be provided by `openshift-install` itself, but it seems to me we should at least leave the door open to shipping something higher level that does aid in automating UPI, somewhat like the [assisted installer for metal](https://github.com/openshift/enhancements/blob/master/enhancements/installer/connected-assisted-installer.md).

Anyways I think we're all agreed let's just target this for UPI.

File: docs/user/vsphere/install_upi.md (hunk 1 above)
Comment (cgwalters) [reply docs/user/vsphere/install_upi.md:261]: Create a new machine, i.e. reprovision.

Comment (lucab) [reply docs/user/vsphere/install_upi.md:261]: This plugs into the same flow as the bare-metal: kargs should get translated into NM configuration in the initramfs, and that is forwarded to the real rootfs.

**Comment by cgwalters** [comment]
OK I noticed there were two existing sections on static IP and hostname; I consolidated them and replaced with this.

However, clearly this would be much more compelling if we updated the existing Terraform code for UPI vSphere with static IPs to use this...I don't speak TF well enough to do that myself.

**Comment by cgwalters** [comment]
> Was static addressing a feature for 4.6?

Yes, it was part of https://github.com/openshift/enhancements/blob/master/enhancements/rhcos/static-networking-enhancements.md

> I only see this possible for 4.6 in UPI with documentation change but I doubt QE has tested it. 

Agree.

> I also think we should be updating the UPI terraform template to replace our current method of addressing (which is different than #3533). And is something we should already be testing in CI for 4.6?

Yes to both!  I know we're somewhat late in the cycle now but this is a big ergonomic and implementation benefit for vSphere UPI.  Would you have a few cycles to look at this?

File: docs/user/vsphere/install_upi.md [hunk 2]
```diff
@@ -279,72 +279,27 @@ The Ignition config created by the OpenShift Installer cannot be used directly b
...
-        "path": "/etc/sysconfig/network-scripts/ifcfg-ens192",
-        "contents": {
-          "source": "data:text/plain;charset=utf-8;base64,VFlQRT1FdGhlcm5ldApCT09UUFJPVE89bm9uZQpOQU1FPWVuczE5MgpERVZJQ0U9ZW5zMTkyCk9OQk9PVD15ZXMKSVBBRERSPTEwLjAuMC4yClBSRUZJWD0yNApHQVRFV0FZPTEwLjAuMC4xCkRPTUFJTj1teWRvbWFpbi5jb20KRE5TMT04LjguOC44",
-        },
-        "mode": 420
-      }
-    ]
-  },
-}
+govc vm.power -on "${VM_NAME}"
```
Comment (sferich888) [inline docs/user/vsphere/install_upi.md:299]: `govc` is a community tool (offered by vsphere). 
Who supports the use/distribution and maintenance of this tool?

Comment (cgwalters) [reply docs/user/vsphere/install_upi.md:299]: I think UPI installations may necessarily involve some tools not shipped by Red Hat.  The goal here is more to outline suggested approaches, not to *require* any specific tooling.  However we probably should have guidance on this.  I don't have expertise in vSphere enough to say, but I'm sure there are alternatives to using `govc` that we could discuss in a separate section.

**Comment by LorbusChris** [comment]
I'm trying what's described here in https://github.com/openshift/installer/pull/3533, but it's still not working.

**Comment by tvanderka** [comment]
FYI this worked for me on OKD:
guestinfo.afterburn.initrd.network-kargs=ip=10.20.35.21::10.20.35.1:24:okd-worker-1::none:10.20.0.11

Creates /etc/NetworkManager/system-connections/default_connection.nmconnection with
[ipv4]
address1=10.20.35.31/24,10.20.35.1
dhcp-hostname=okd-worker-1
dns=10.20.0.11;
dns-search=

Works fine, even though no dns-search/domain is set so short name resolution does not work. And interface=ens192 may be a good idea if using multiple interfaces.

**Pull request closed without merging**